language/compiler/ir-to-bytecode/syntax/src/lexer.rs

```diff
@@ -311,41 +132,248 @@ impl<'input, 'builder> Lexer<'input, 'builder> {
+            }
+            '=' => {
+                if text.starts_with("==") {
+                    (Tok::EqualEqual, 2)
+                } else {
+                    (Tok::Equal, 1)
+                }
+            }
+            '!' => {
+                if text.starts_with("!=") {
+                    (Tok::ExclaimEqual, 2)
+                } else {
+                    (Tok::Exclaim, 1)
                 }
             }
-            match index {
-                None => Err(ParseError::InvalidToken {
+            '<' => {
+                if text.starts_with("<=") {
+                    (Tok::LessEqual, 2)
+                } else {
+                    (Tok::Less, 1)
+                }
+            }
+            '>' => {
+                if text.starts_with(">=") {
+                    (Tok::GreaterEqual, 2)
+                } else {
+                    (Tok::Greater, 1)
+                }
+            }
+            '%' => (Tok::Percent, 1),
+            '(' => (Tok::LParen, 1),
+            ')' => (Tok::RParen, 1),
+            '*' => (Tok::Star, 1),
+            '+' => (Tok::Plus, 1),
+            ',' => (Tok::Comma, 1),
+            '-' => (Tok::Minus, 1),
+            '.' => (Tok::Period, 1),
+            '/' => (Tok::Slash, 1),
+            ':' => (Tok::Colon, 1),
+            ';' => (Tok::Semicolon, 1),
+            '^' => (Tok::Caret, 1),
+            '{' => (Tok::LBrace, 1),
+            '}' => (Tok::RBrace, 1),
+            _ => {
+                return Err(ParseError::InvalidToken {
                     location: start_offset,
-                }),
-                Some(index) => {
-                    let result = &text[..longest_match];
-                    let remaining = &text[longest_match..];
-                    let end_offset = start_offset + longest_match;
-                    self.text = remaining;
-                    self.consumed = end_offset;
-                    self.token = (
-                        start_offset,
-                        Token(Tok::from_usize(index), result),
-                        end_offset,
-                    );
-                    Ok(())
+                });
+            }
+        };
+
+        let result = &text[..len];
+        let remaining = &text[len..];
+        let end_offset = start_offset + len;
+        self.text = remaining;
+        self.consumed = end_offset;
+        self.token = (start_offset, Token(tok, result), end_offset);
+        Ok(())
+    }
+}
+
+// Return the length of the substring matching [a-zA-Z$_][a-zA-Z0-9$_]
+fn get_name_len(text: &str) -> usize {
+    let mut len = 0;
+    for c in text.chars() {
+        match c {
+            'a'..='z' | 'A'..='Z' | '$' | '_' => {
+                len += 1;
+            }
+            '0'..='9' => {
+                if len == 0 {
+                    break;
                 }
+                len += 1;
+            }
+            _ => {
+                break;
             }
         }
     }
+    len
```

Comment:
```
if text.starts_with('0'..='9') { return 0 } // I don't think this works but you get the idea
text.chars().take_while(|c| match c { 'a'..='z' | 'A'..='Z' | '$' | '_'  | '0'..='9' => true | _ => false }).size_hint().1.unwrap()
````
or 
```
if text.starts_with('0'..='9') { return 0 } // I don't think this works but you get the idea
text.chars().take_while(|c| match c { 'a'..='z' | 'A'..='Z' | '$' | '_'  | '0'..='9' => true | _ => false }).map(|_| 1).sum()
```

Reply:
For the first line, how about the following?
`if ('0'..='9').contains(&text.chars().next().unwrap_or('0')) { return 0; }`

Reply:
I split that first part into 2 lines and then used a suggestion from Tim to use the iterator's position() function to implement the second part (instead of take_while()).